Add support for returning documents with completion suggester

areek opened the pull request:
This is a followup to https://github.com/elastic/elasticsearch/pull/13576#issuecomment-140695698, enriching completion suggestions
 with documents.

This commit enables completion suggester to return documents
associated with suggestions. Now the document source is returned
with every suggestion, which respects [source filtering options](https://www.elastic.co/guide/en/elasticsearch/reference/master/search-request-source-filtering.html).

In case of suggest queries spanning more than one shard, the
suggest is executed in two phases, where the last phase fetches
the relevant documents from shards, implying executing suggest
requests against a single shard is more performant due to the
document fetch overhead when the suggest spans multiple shards.

Example completion suggest response:
```bash
{
  "song-suggest": [
    {
      "text": "nev",
      "offset": 0,
      "length": 4,
      "options": [ {
          "text": "Nevermind",
          "_index": "music",
          "_type": "song",
          "_id": "52",
          "_score": 52,
          "_source": {
              "song": "Nevermind",
              "artist": "Nirvana"
          }
        }
      ]
    }
  ]
}
```
NOTE: now that we support returning documents with suggestions, we can remove the `payload` option

relates https://github.com/elastic/elasticsearch/issues/10746

areek commented on core/src/main/java/org/elasticsearch/search/suggest/Suggest.java:
```diff
@@ -160,18 +182,6 @@ public static Suggest readSuggest(StreamInput in) throws IOException {
-        return groupedSuggestions;
-    }
-
     public static List<Suggestion<? extends Entry<? extends Option>>> reduce(Map<String, List<Suggest.Suggestion>> groupedSuggestions) {
```
we should extract the reduce logic from suggest, the suggest abstraction is not the right place for these, maybe somewhere in `SearchPhraseController`? I think this will simplify `Suggest` and help with nuking the crazy generics. maybe we should simplify this in a subsequent PR?

s1monw replied:
++

s1monw commented:
@areek thanks so much! It took me a while to figure out what you did with the named suggestions etc. I do think there is an easier way to do what you wanted to do. I think we can, instead of use a `name, scoredoc[]` map we can just use offsets into the `ScoreDoc[]` which would allow us to keep most of the code as is and just resolve the actual documents once we finish up after the fact. ie. query would be implicitly using offset 0 and each of the suggestion will get an offset assigned in the beginning since we know the size of each suggestion and the query ahead of time. 
The code that actually fetches the docs can still simply be a single array so that code can stay! We might need to add some offset, lenght parameters to stuff like `#fillDocIdsToLoad` but I think that is less intrusive than our current path?

I love the tests man thanks for that

rmuir commented:
Can we improve the abstractions here? For example, the relatively straightforward `ScoreDoc[]` and `IntArrayList` get hidden behind less-obvious layers named `SortedDocs` and `DocIdsToLoad`, and neither of these have any javadocs. It makes it hard to tell what is going on with the code.

areek commented:
@rmuir I am working on removing the `DocIdsToLoad` abstraction and changing the  `SortedDocs` to store offsets instead of map of `IntArrayList`, as per @s1monw's  comment. Will make sure to add javadocs, to clarify. Thanks for the feedback :)

areek commented:
@s1monw @rmuir I updated the PR to represent search and suggest score docs array via an offsets array in `SortedDocs` abstraction. Would love to get feedback on the current approach!

s1monw commented on core/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:
```diff
@@ -74,7 +76,7 @@
     private volatile AtomicArray<ShardSearchFailure> shardFailures;
     private final Object shardFailuresMutex = new Object();
-    protected volatile ScoreDoc[] sortedShardList;
+    protected volatile SortedDocs sortedShardList;
```
can we rename this to `sortedShardDocs`

s1monw commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -321,7 +324,8 @@ public QuerySearchResult executeQueryPhase(QuerySearchRequest request) {
             long time = System.nanoTime();
             queryPhase.execute(context);
-            if (context.queryResult().topDocs().scoreDocs.length == 0 && context.scrollContext() == null) {
+            if (context.queryResult().topDocs().scoreDocs.length == 0 && context.scrollContext() == null
```
maybe we can factor out a method `boolean hasHits(QuerySearchResult result)` it's used in two places an a complex condition

s1monw commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -812,40 +816,48 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
+        }
+        int[] docIdsToLoad = searchDocIdsToLoad;
+        final Suggest suggest = context.queryResult().suggest();
+        if (suggest != null && suggest.hasScoreDocs()) {
```
I think what would make this simpler is if we had an API on `List<CompletionSuggestion> suggestions = suggest.filter(CompletionSuggestion.class)` that we can use to iterate over twice:
```
int numSuggestDocs = 0;
for (CompletionSuggestion completionSuggestion : suggestions) {
  numSuggestDocs += completionSuggestion.getScoreDocs().length
}

// now fill the docIDs from the search and use the `numSuggestDocs` to extend the size of the array.
```
makes sense?

areek replied:
Thanks for the suggestion, it did simplify logic to extend the doc array for suggest docs

s1monw commented on core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java:
```diff
@@ -258,20 +303,53 @@ private static long optionalSum(long left, long right) {
+    public ScoreDoc[] getLastEmittedDocPerShard(SortedDocs sortedDocs, int numShards) {
         ScoreDoc[] lastEmittedDocPerShard = new ScoreDoc[numShards];
-        for (ScoreDoc scoreDoc : sortedShardList) {
+        for (int sortedDocsIndex = sortedDocs.getOffset(0); sortedDocsIndex < sortedDocs.getOffset(1); sortedDocsIndex++) {
```
`sortedDocs.getOffset(0)` is always == 0 right?

s1monw replied:
if so, I think we should have a signature like: `public ScoreDoc[] getLastEmittedDocPerShard(ScoreDocs[] sortedShardList, int numShards, int offset, int length) {` it's more flexible I think

s1monw commented on core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java:
```diff
@@ -349,7 +427,8 @@ public InternalSearchResponse merge(ScoreDoc[] sortedDocs, AtomicArray<? extends
         List<InternalSearchHit> hits = new ArrayList<>();
         if (!fetchResults.isEmpty()) {
-            for (ScoreDoc shardDoc : sortedDocs) {
+            for (int i = sortedDocs.getOffset(0); i < sortedDocs.getOffset(1); i++) {
```
also this one is implicit `int i = 0;`? and maybe we can also derive the length from the request? if we can to that, I think we can get rid of the SortedDocs class entirely and just use implicit offsets from the request length / size?

s1monw commented on core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java:
```diff
@@ -360,54 +439,79 @@ public InternalSearchResponse merge(ScoreDoc[] sortedDocs, AtomicArray<? extends
+                        if (suggestion instanceof CompletionSuggestion) {
+                            CompletionSuggestion.Entry entry = (CompletionSuggestion.Entry) suggestion.getEntries().get(0);
+                            final List<CompletionSuggestion.Entry.Option> suggestionOptions = entry.getOptions();
+                            int nextOffset = sortedDocs.getOffset(i + 1);
```
can we derive them from the request length? I think it should be possible?

s1monw commented on core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java:
```diff
@@ -360,54 +439,79 @@ public InternalSearchResponse merge(ScoreDoc[] sortedDocs, AtomicArray<? extends
-                    continue;
+        if (firstResult.suggest() != null) {
+            final Map<String, List<Suggestion>> groupedSuggestions = new HashMap<>();
+            queryResults.stream()
```
can we implement this in a non-functional way? I have a hard time to understand that

s1monw commented on core/src/main/java/org/elasticsearch/search/fetch/matchedqueries/MatchedQueriesFetchSubPhase.java:
```diff
@@ -43,7 +43,7 @@
     @Override
     public void hitsExecute(SearchContext context, InternalSearchHit[] hits) {
-        if (hits.length == 0) {
+        if (hits.length == 0 || context.parsedQuery() == null) {
```
hmm is that true? i wonder if this is equivalent to a match all? I could be very wrong.

areek replied:
this is not equivalent to a match all, the parsedQuery is null only when the search request only has suggest and we skip the search context preprocessing [here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/search/internal/DefaultSearchContext.java#L192), ensuring we don't run a match all query when a search request only has suggest and no explicit search query

s1monw replied:
please add this as an inline comment to this.

s1monw commented on core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java:
```diff
@@ -415,8 +415,7 @@ public void setInnerHits(Map<String, InternalSearchHits> innerHits) {
 
-    @Override
-    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+    public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) throws IOException {
```
why did you do this? I am curious :) it can also be private or protected?

areek replied:
`CompletionSuggestion.Entry.Option#innerToXContent` uses the `toInnerXContent` for rendering the hit as part of the option, instead of rendering the hit as a seperate object.

s1monw replied:
fair enough - can you leave a comment

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/Suggest.java:
```diff
@@ -97,6 +95,16 @@ public int size() {
+    /**
+     * Whether any suggestions had query hits
+     */
+    public boolean hasScoreDocs() {
```
can we compute this only once at creation time?

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/Suggest.java:
```diff
@@ -68,6 +65,7 @@ public Suggest() {
     }
 
     public Suggest(List<Suggestion<? extends Entry<? extends Option>>> suggestions) {
+        suggestions.sort((o1, o2) -> o1.getName().compareTo(o2.getName()));
```
any reason why we sort here? can you add a comment

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -62,6 +69,24 @@ public CompletionSuggestion(String name, int size) {
+        if (hasScoreDocs() == false) {
+            return new ScoreDoc[0];
+        }
+        assert getEntries().size() == 1 : "CompletionSuggestion must have only one entry";
```
can it have 0?

s1monw replied:
ah nevermind I see you do the `hasScoreDocs` thing

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -90,31 +115,50 @@ protected boolean lessThan(Entry.Option a, Entry.Option b) {
+    public void setShardIndex(int shardIndex) {
+        super.setShardIndex(shardIndex);
+        assert entries.size() == 1 : "CompletionSuggestion must have only one entry";
+        entries.get(0).forEach(option ->  option.setShardIndex(shardIndex));
```
this only works if we have entries? shall we check? Also can we just do a for loop here instead

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -171,14 +218,30 @@ protected void mergeInto(Suggest.Suggestion.Entry.Option otherOption) {
+                this.doc.shardIndex = shardIndex;
+            }
+
+            public void hit(InternalSearchHit hit) {
```
maybe call this `setHit`

s1monw commented:
@areek this looks pretty awesome already I left some comments

areek commented:
@s1monw Thanks for the review! I updated the PR, addressing all your feedback. After some thinking, I managed to remove the `SortedDocs` abstraction all together, now we just use `ScoreDoc[]` for both search and suggest docs. Would be awesome if you could review it again?

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchDfsQueryThenFetchAsyncAction.java:
```diff
@@ -196,7 +195,8 @@ private void finishHim() {
+                final InternalSearchResponse internalResponse = searchPhaseController.merge(useScroll, sortedShardDocs, queryResults,
                     fetchResults);
                 String scrollId = null;
                 if (request.scroll() != null) {
```
can we use `isScroll` here again?

s1monw commented on core/src/main/java/org/elasticsearch/action/search/SearchDfsQueryThenFetchAsyncAction.java:
```diff
@@ -196,7 +195,8 @@ private void finishHim() {
             @Override
             public void doRun() throws IOException {
-                final InternalSearchResponse internalResponse = searchPhaseController.merge(sortedShardList, queryResults,
+                boolean useScroll = request.scroll() != null;
```
should this be final and maybe called `isScroll` or `isScrollRequest`

areek replied:
changed this and in Search*AsyncAction to `isScrollRequest`

s1monw commented on core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java:
```diff
@@ -191,19 +193,39 @@ private static long optionalSum(long left, long right) {
+            Suggest suggest = result.queryResult().suggest();
+            if (suggest != null) {
+                int numSuggestDocs = 0;
+                for (CompletionSuggestion suggestion : suggest.filter(CompletionSuggestion.class)) {
```
please assign `suggest.filter(CompletionSuggestion.class)` to a local var so we can reuse it below when we iterate it a second time. I also wonder if we should do this ahead of time such that we can incorporate the suggest size into the creation of `docs` so we don't need to copy it?

areek replied:
oops, thanks for the suggestion. I think the restructuring made the code simpler

s1monw commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -811,40 +814,53 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
+        final Suggest suggest = context.queryResult().suggest();
+        if (suggest != null && suggest.hasScoreDocs()) {
+            int numSuggestDocs = 0;
+            for (CompletionSuggestion completionSuggestion : suggest.filter(CompletionSuggestion.class)) {
```
please assign suggest.filter(CompletionSuggestion.class) to a local var so we can reuse it below when we iterate it a second time.

s1monw replied:
I also wonder if we should move this part in a different method such that this method doesn't get too big to be inlined if needed but I guess it's not hot enough anyhow.

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -62,6 +69,23 @@ public CompletionSuggestion(String name, int size) {
         super(name, size);
     }
 
+    public ScoreDoc[] getScoreDocs() {
```
can this just be an `Iterable<ScoreDoc>` I think it's only used as a such so we don't need to make a copy of it.

areek replied:
Thanks for the suggestion.

s1monw commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -811,40 +814,53 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
+            docIdsToLoad = new int[searchDocIdsToLoad.length + numSuggestDocs];
+            System.arraycopy(searchDocIdsToLoad, 0, docIdsToLoad, 0, searchDocIdsToLoad.length);
+            int offset = searchDocIdsToLoad.length;
+            for (CompletionSuggestion completionSuggestion : suggest.filter(CompletionSuggestion.class)) {
```
this code and the code in `SearchPhaseController#sortDocs` is almost identical. I think the only difference is the assignment of the shard ID. Maybe we can factor this out into a static method and use it in both places. It would be good to redcue the duplication on such a level and it would increase the test coverage. I would be in favor of that.

areek replied:
do you mean to share the code between the single shard case in`SearchPhaseController#sortedDocs` and this?

s1monw commented on core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java:
```diff
@@ -258,40 +274,79 @@ private static long optionalSum(long left, long right) {
-            return getLastEmittedDocPerShard(sortedShardList, numShards);
-        } else {
-            return null;
+        ScoreDoc[] scoreDocs = mergedTopDocs.scoreDocs;
```
as far as I can tell this is only relevant in a scroll context. I am not 100% following why you implemented this here?

s1monw commented on core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java:
```diff
@@ -436,4 +517,14 @@ public InternalSearchResponse merge(ScoreDoc[] sortedDocs, AtomicArray<? extends
         return new InternalSearchResponse(searchHits, aggregations, suggest, shardResults, timedOut, terminatedEarly);
     }
 
+    private static int topN(List<? extends AtomicArray.Entry<? extends QuerySearchResultProvider>> queryResults) {
```
can you document this method  it's not obvious what it does.

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/Suggest.java:
```diff
@@ -68,7 +67,12 @@ public Suggest() {
+        // we sort suggestions by their names to ensure iteration over suggestions are consistent
+        // this is needed as we need to fill in suggestion docs in SearchPhaseController#sortDocs
+        // in the same order as we enrich the suggestions with fetch results in SearchPhaseController#merge
+        suggestions.sort((o1, o2) -> o1.getName().compareTo(o2.getName()));
```
I am confused why do we need to sort? Is this due to suggestions from different shards can come back in different order?

areek replied:
we need to sort because we use a `Map<String, List<Suggestion>>` to group suggestions when we merge suggestions in `SearchPhaseController#merge` and when we reduce suggestions in `SearchPhaseController#sortDocs`, so there is no guarantee that the order of suggestions in both cases would be consistent.

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/Suggest.java:
```diff
@@ -238,6 +245,10 @@ public String getName() {
             return name;
         }
 
+        public int getSize() {
```
add a javadoc for this please

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -90,30 +114,51 @@ protected boolean lessThan(Entry.Option a, Entry.Option b) {
+
+    @Override
+    public Suggest.Suggestion<Entry> reduce(List<Suggest.Suggestion<Entry>> toReduce) {
+        return reduceTo(new ArrayList<>(toReduce));
```
why do you make a copy here? is it modified?

areek replied:
this was my naive attempt to keep the generics happy, fixed

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -145,9 +190,12 @@ protected Option newOption() {
         public static class Option extends Suggest.Suggestion.Entry.Option {
             private Map<String, Set<CharSequence>> contexts;
             private Map<String, List<Object>> payload;
+            private ScoreDoc doc;
```
can this be final?

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -35,6 +40,8 @@
 import java.util.Map;
 import java.util.Set;
 
+import static org.elasticsearch.search.suggest.Suggest.COMPARATOR;
```
the generic here are aweful. I regret it so much that I added them. We should really clean this up it makes stuff so complicated for no good reason

areek replied:
++, I will try to clean up the generics in `Suggest` in subsequent PRs

s1monw commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -90,30 +114,51 @@ protected boolean lessThan(Entry.Option a, Entry.Option b) {
-    public Suggest.Suggestion<Entry> reduce(List<Suggest.Suggestion<Entry>> toReduce) {
-        if (toReduce.size() == 1) {
-            return toReduce.get(0);
+    public static CompletionSuggestion reduceTo(List<Suggest.Suggestion> toReduce) {
```
I am not sure we do have that but if not can we have a test for this method as well as javadocs?

s1monw commented:
this looks awesome @areek it's simplified so much from the original patch. All the extra classes and abstractions are gone! good job! I added some comments about sharing more code etc. I also think we should have more tests on the CompletionSuggest end, like simple unittests of filter methods etc. 
What I am really missing is a place where we document how the fetching works, that the order of the docIDs is crucial. I think we should add it where we create the docId array?  I think we are super close here, can you remove the WIP label?

s1monw commented:
@clintongormley can you please check if we need to work more on docs here?

areek commented:
Thanks @s1monw for the review! I have added some documentation about how fetching works in `SearchPhaseController#sortedDocs` and what is expected by `SearchPhaseController#merge`, added simple unittests for `Suggest` and `CompletionSuggester` (more tests can be added later?).

s1monw commented:
LGTM - I think CI failed so I guess you want to merge master in again and run it again! good job.

areek commented:
@clintongormley if you any feedback regarding the docs, let me know, will do a separate PR for it.

speedplane commented:
I'm using ES 1.7 now and the completion suggester is currently one of my largest memory bottlenecks, so thank you, I'm definitely looking forward to this! Is there documentation yet that describes this new feature? 

In my index, to save space, I do not have a _store, so if I want to use this feature, my understanding is that: 

1. I'll have to create a new index that has _store turned on.
2. During indexing, I put the payloads into this new index. Their ID can be keyed off of their value.
3. Also when indexing, I point the completion suggester to the documents in the new index.

Do I have this generally correct?

trompx commented:
Hello,

Compared to the old completion suggester, do we still need to optimize indices to prevent duplicate results when we update some documents ?

As quoted here : https://discuss.elastic.co/t/autocompletion-suggester-duplicate-record-in-suggestion-return/16950/4

> The main reason, why you see the suggestion twice is, that even though a 
document is deleted and cannot be found anymore, the suggest data 
structures are only cleaned up during merges/optimizations. Running 
optimize should fix this.

Anyway awesome work on this new suggester. By any chance have you run any benchmark to compare suggestion speed 2.2 vs 5.0?

areek commented:
Hey @trompx,

> Compared to the old completion suggester, do we still need to optimize indices to prevent duplicate results when we update some documents ?

No we don't need to as the new completion suggester is near-real time and is expected to take into account deleted documents (even if the deleted document hasn't been merged away).

 > Anyway awesome work on this new suggester. By any chance have you run any benchmark to compare suggestion speed 2.2 vs 5.0?

Thanks :), the benchmark for the new implementation can be found [here](https://github.com/elastic/elasticsearch/issues/10746#issuecomment-157276138)
